org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/semantictokens/SemanticTokensService.java

```diff
@@ -0,0 +1,205 @@
+			final TextDocumentIdentifier textDocumentIdentifier, final CancelIndicator cancelIndicator) {
+		URI uri = getURI(textDocumentIdentifier);
+		ISemanticHighlightingCalculator semanticHighlightingCalculator = getService(uri,
+				ISemanticHighlightingCalculator.class);
+		MergingHighlightedPositionAcceptor acceptor = new MergingHighlightedPositionAcceptor(
+				semanticHighlightingCalculator);
+		semanticHighlightingCalculator.provideHighlightingFor(resource, acceptor, cancelIndicator);
+		return acceptor.getPositions();
+	}
+
+	/**
+	 * Semantic tokens full.
+	 *
+	 * @param params
+	 *            the params
+	 * @param cancelIndicator
+	 *            the cancel indicator
+	 * @return the semantic tokens
+	 */
+	public SemanticTokens semanticTokensFull(final Document document, final XtextResource resource,
+			final SemanticTokensParams params, final CancelIndicator cancelIndicator) {
+		List<Integer> data = new ArrayList<>();
+		int lastOffset = 0;
+		int lastLine = 0;
+		for (LightweightPosition lightweihtPosition : getPositions(resource, params.getTextDocument(),
+				cancelIndicator)) {
+			Integer tokenType = getTokenType(lightweihtPosition.getIds());
+			if (tokenType != -1) {
+				int deltaOffset = lightweihtPosition.getOffset() - lastOffset;
+				// TODO: create a getPosition with state, it is very suboptimal that this method will always start from
+				// scratch
+				// and we now the positions are sorted, so we should just start from where we left and add the result
+				// to the previous one
+				Position position = document.getPosition(lightweihtPosition.getOffset());
+				int deltaLine = position.getLine() - lastLine;
+				data.add(deltaLine); // delta line
+				lastLine = position.getLine();
+				data.add(deltaLine == 0 ? deltaOffset : position.getCharacter()); // delta start relative to previous
+																					// token if on the same line or to 0
+				lastOffset = lightweihtPosition.getOffset();
+				data.add(lightweihtPosition.getLength()); // length
+				data.add(tokenType); // token type
+				data.add(0); // token modifiers
+			}
+		}
+		return new SemanticTokens(data); // resultId not set, no delta support
+	}
+
+	/**
+	 * Semantic tokens range.
+	 *
+	 * @param params
+	 *            the params
+	 * @param cancelIndicator
+	 *            the cancel indicator
+	 * @return the semantic tokens
+	 */
+	public SemanticTokens semanticTokensRange(final Document document, final XtextResource resource,
+			final SemanticTokensRangeParams params, final CancelIndicator cancelIndicator) {
+		List<Integer> data = new ArrayList<>();
+		Position rangeStart = params.getRange().getStart();
+		Position rangeEnd = params.getRange().getEnd();
+		int lastOffset = 0;
+		int lastLine = 0;
+		for (LightweightPosition lightweihtPosition : getPositions(resource, params.getTextDocument(),
+				cancelIndicator)) {
+			// TODO: it would be nicer to filter before calculating the list of LightweightPosition, to make the
+			// computation faster
```

Comment:
any idea how to implement the semantic highlight calculator to only calculate highlights for a given range? Is there already an utility go get/iterate through the nodes of a given offset and length?

Reply:
Hi @cdietrich , any input on how to improve the performance of this part? That would be very welcome.

Reply:
I think an approach based on the line breaks and binary search might be fruitful to achieve decent performance, e.g. similar what needs to be done for the full list of semantic tokens, but on a subrange of the offsets / sorted positions.

Reply:
I am thinking rather something along this line:
Extend IHighlightedPositionAcceptor acceptor with a method accepts(offset, length), and then in 
org.eclipse.xtext.ide.editor.syntaxcoloring.DefaultSemanticHighlightingCalculator.highlightElementRecursively(EObject, IHighlightedPositionAcceptor, CancelIndicator) do
```
	protected void highlightElementRecursively(EObject element, IHighlightedPositionAcceptor acceptor,
			CancelIndicator cancelIndicator) {
		TreeIterator<EObject> iterator = EcoreUtil2.eAll(element);
		while (iterator.hasNext()) {
			EObject object = iterator.next();
                        INode node = NodeModelUtils.genNode(object);
			if (acceptor.acceps(node.getOffset(), node.getLength()) && highlightElement(object, acceptor, cancelIndicator)) {
				iterator.prune();
			}
		}
	}
```

I added the `INode node = NodeModelUtils.genNode(object);` and `acceptor.acceps(node.getOffset(), node.getLength())`. We could call the method different, but the idea would be the same, not to color parts of the AST which are outside of the range.

I assume this would be faster than coloring them and then filtering them out. What do you think?

Reply:
> Extend IHighlightedPositionAcceptor acceptor with a method accepts(offset, length)

Instead of extending the interface, please rather use a decorator that filters the invalid / out-of-range positions and passes the "correct" positions into another acceptor.